File: scripts/install/src/cq_install/hosts/windsurf.py
"""Windsurf host adapter."""

from __future__ import annotations

from pathlib import Path

from cq_install.common import (
    copy_tree,
    remove_copied_tree,
    remove_json_entry,
    upsert_json_entry,
)
from cq_install.content import CQ_MCP_KEY, cq_binary_name
from cq_install.context import ChangeResult, InstallContext
from cq_install.hosts.base import HostDef
from cq_install.runtime import runtime_root

WINDSURF_HOST_SKILLS_MANIFEST = ".cq-install-manifest.json"
WINDSURF_MCP_FILE = "mcp_config.json"
WINDSURF_MCP_SERVERS_KEY = "mcpServers"
WINDSURF_SKILLS_DIR = "skills"

# Windsurf stores its config under ~/.codeium/windsurf/ on every platform.
# Confirmed against a live Windows 11 install.
WINDSURF_GLOBAL_TARGET = Path(".codeium") / "windsurf"


class WindsurfHost(HostDef):
    """Adapter for the Windsurf editor host.

    Windsurf stores its config under `~/.codeium/windsurf/` on every
    platform. Confirmed against a live Windows 11 install where the
    directory contains Cascade/brain/memories subdirs. `Path.home()`
    resolves to `%USERPROFILE%` on Windows, so this works without any
    platform branching.
    """

    name = "windsurf"
    supports_project = False

    def global_target(self) -> Path:
        """Return the global Windsurf config dir."""
        return Path.home() / WINDSURF_GLOBAL_TARGET

    def install(self, ctx: InstallContext) -> list[ChangeResult]:
        """Install cq into the Windsurf target."""
        results: list[ChangeResult] = []
        if ctx.host_isolated_skills:
            results.append(
                copy_tree(
                    ctx.plugin_root / WINDSURF_SKILLS_DIR,
                    ctx.target / WINDSURF_SKILLS_DIR,
                    manifest_name=WINDSURF_HOST_SKILLS_MANIFEST,
                    dry_run=ctx.dry_run,
                )
            )
        else:
            results.extend(ctx.run_state.ensure_shared_skills(ctx))
        results.extend(ctx.run_state.ensure_cq_binary(ctx))
        binary_path = runtime_root() / "bin" / cq_binary_name()
        results.append(
            upsert_json_entry(
                ctx.target / WINDSURF_MCP_FILE,
                [WINDSURF_MCP_SERVERS_KEY, CQ_MCP_KEY],
                {
                    "command": str(binary_path),
                    "args": ["mcp"],
                },
                dry_run=ctx.dry_run,
            )
        )
        return results

    def uninstall(self, ctx: InstallContext) -> list[ChangeResult]:
        """Remove cq from the Windsurf target."""
        results: list[ChangeResult] = [
            remove_copied_tree(
                ctx.target / WINDSURF_SKILLS_DIR,
                manifest_name=WINDSURF_HOST_SKILLS_MANIFEST,
                dry_run=ctx.dry_run,
            ),
            remove_json_entry(
                ctx.target / WINDSURF_MCP_FILE,
                [WINDSURF_MCP_SERVERS_KEY, CQ_MCP_KEY],
                dry_run=ctx.dry_run,
            ),
        ]
        return results

File: cli/internal/install/windsurf.go
package install

import (
	"path/filepath"

	"github.com/mozilla-ai/cq/sdk/go/prompts"
)

// windsurfMCPFile is the MCP configuration file Windsurf reads on every platform.
const windsurfMCPFile = "mcp_config.json"

// windsurfHost installs cq into the Windsurf editor (skill commons + MCP entry).
//
// Windsurf stores its config under ~/.codeium/windsurf on every platform and
// is global-only; it reads skills from the shared commons.
type windsurfHost struct{}

// init registers the Windsurf adapter so SelectHosts can return it.
func init() {
	registry["windsurf"] = windsurfHost{}
}

// GlobalTarget returns the Windsurf config dir under home.
func (windsurfHost) GlobalTarget(home string) string {
	return windsurfTarget(home)
}

// Install writes the shared skill and the cq MCP server entry.
func (windsurfHost) Install(ctx Context) ([]Change, error) {
	skill, err := writeManagedFiles(ctx.SkillsDir, map[string]string{
		filepath.Join("cq", "SKILL.md"): prompts.Skill(),
	}, ctx.DryRun)
	if err != nil {
		return nil, err
	}
	mcp, err := upsertJSONEntry(
		filepath.Join(ctx.Target, windsurfMCPFile),
		[]string{"mcpServers", "cq"},
		map[string]any{"command": ctx.BinaryPath, "args": []any{"mcp"}},
		ctx.DryRun,
	)
	if err != nil {
		return nil, err
	}
	return []Change{skill, mcp}, nil
}

// Name returns the host identifier.
func (windsurfHost) Name() string { return "windsurf" }

// SupportsProject reports that Windsurf is global-only.
func (windsurfHost) SupportsProject() bool { return false }

// Uninstall removes the cq MCP entry.
//
// NOTE: the skill lives in the shared commons (~/.agents/skills), which other
// hosts may also use, so uninstalling one host must not remove it; the shared
// skill is intentionally left in place.
func (windsurfHost) Uninstall(ctx Context) ([]Change, error) {
	mcp, err := removeJSONEntry(
		filepath.Join(ctx.Target, windsurfMCPFile),
		[]string{"mcpServers", "cq"},
		ctx.DryRun,
	)
	if err != nil {
		return nil, err
	}
	return []Change{mcp}, nil
}
